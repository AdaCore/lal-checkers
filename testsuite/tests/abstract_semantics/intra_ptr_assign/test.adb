procedure Test is
   type Point is record
      x : Integer;
      y : Integer;
   end record;

   type Rectangle is record
      center : Point;
      width : Integer;
      height : Integer;
   end record;

   p : access Integer;
   r : Rectangle;
begin
   p := r.center.x'Access;
   if p.all = 2 then
      r.width := p.all + 2;
   end if;
end Ex1;
