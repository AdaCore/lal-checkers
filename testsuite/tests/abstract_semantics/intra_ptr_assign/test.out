{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assume0",
          "assume1",
          "assume2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [4, 4], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume1",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume2": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume1",
          "assume2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume3": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume3",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([-2147483648, 2147483647], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{[all-path]}"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assume0",
          "assume1",
          "assume2",
          "assume3",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([-2147483648, 2147483647], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{[all-path]}"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assume0",
          "assume1",
          "assume2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [4, 4], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume1",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume2": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume1",
          "assume2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "assume3": [
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume3",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([-2147483648, 2147483647], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{[all-path]}"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assume0",
          "assume1",
          "assume2",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([2, 2], [-2147483648, 2147483647]), [4, 4], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      },
      {
        "trace:": [
          "assign0",
          "assume0",
          "assume3",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '(([-2147483648, 2147483647], [-2147483648, 2147483647]), [-2147483648, 2147483647], [-2147483648, 2147483647])'}, 0)",
          "p": "{Get_0(Get_0(0x00000002))}"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "p": "{[all-path]}"
        }
      }
    ]
  }
}